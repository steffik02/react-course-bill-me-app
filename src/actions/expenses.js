import uuid from 'uuid';
import database from '../firebase/firebase';

        //use push
        //attach then callback
        //dispatch action
        //redirect

// add expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = { description, note, amount, createdAt }
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        })
    }
}

// remove expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//create new action with same call signature as removeExpense
//test startRemoveExpense with 'should remove expenses from firebase'
    // to test this you can just try to fetch it, call val on the snapshot, and if no data null will be return value
//use startRemoveExpense in EditExpensePage instead of removeExpense
//Adjust EditExpensePage tests
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({id}));
        });
    };
};

// edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// fetch all expense data from the db once
// parse the data into an array (see firebase.js file)
// dispatch SET_EXPENSES to change the data in the store
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
};
