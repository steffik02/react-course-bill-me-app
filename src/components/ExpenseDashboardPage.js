import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        This dashboard will help you steer your expense spaceship into financial freedom.
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;