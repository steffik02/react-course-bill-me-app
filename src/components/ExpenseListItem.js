import React from 'react';
import { Link } from "react-router-dom";

//connect component to access dispatch
//import correct delete functin
//wire up onClick

export default ({id, description, amount, createdAt}) => (
    <div>
        <Link to={'/edit/' + id}>
            <h3>Description: {description}</h3>
        </Link>
        <p>Amount: ${amount/100}</p>
        <p>Date Created: {createdAt}</p>
    </div>
);