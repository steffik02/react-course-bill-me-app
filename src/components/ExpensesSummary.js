import React from 'react';
import {connect} from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses'
import numeral from 'numeral';


export class ExpensesSummary extends React.Component {
    render () {
        const expenseCount = this.props.selectedExpenses.length;
        const expensesTotal = selectExpensesTotal(this.props.selectedExpenses);
        const expensesTotalFormatted = numeral(expensesTotal / 100).format('$0,0.00');
        const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
            return (
            <div className="page-header">
                <div className="content-container">
                    <div>
                        <h1 className="page-header__title">
                            Viewing <span>{expenseCount}</span> {expenseWord} totaling 
                            <span> {expensesTotalFormatted}</span>
                        </h1>
                        <div className="page-header__actions">
                            <Link className="button" to="/create">Add Expense</Link>
                        </div>
                    </div>    
                </div>    
            </div>
)}}
        
        
const mapStateToProps = (state) => ({
    selectedExpenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);