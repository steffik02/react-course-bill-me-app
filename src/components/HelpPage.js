import React from 'react';

const HelpPage = () => (
    <div>
        This is where you get some help. Everybody needs some help sometimes.
    </div>
);

export default HelpPage;