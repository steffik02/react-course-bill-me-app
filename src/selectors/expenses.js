import moment from 'moment';

// Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; 
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const lcdesc = expense.description.toLowerCase();
        const lctext = text.toLowerCase();
        const textMatch = lcdesc.includes(lctext);
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        };
    });
};