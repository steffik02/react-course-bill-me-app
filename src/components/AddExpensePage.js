import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        this.props.history.push('/');
    }
    render () {
        return (
            <div>
                <h1>Add Expense</h1>
                <p>This is where you add expenses. Expenses are bad. Don't add too many.</p>
                <ExpenseForm
                    onSubmit={
                        this.onSubmit
                    }
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);