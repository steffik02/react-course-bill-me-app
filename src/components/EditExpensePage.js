import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// Refactor page to be class based component, pull out inline functions to methods 
// Set up mapDispatchToProps, we need 2 things on the returned object, not just 1: editExpense and removeExpense

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        //dispatch the action to edit the expense
        // props.dispatch(editExpense(props.expense.id, expense));
        this.props.startEditExpense(this.props.expense.id, expense);
        //redirect to the dashboard
        this.props.history.push('/');
    }
    onClick = () => {
        // props.dispatch(removeExpense({ id: props.expense.id }));
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render () {
        return (
            <div>
                <p>This is where you edit your expenses. I will be editing mine to ZERO!</p>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>
                    Remove
                </button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);