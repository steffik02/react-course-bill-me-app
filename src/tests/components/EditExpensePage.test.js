import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

//write test cases: 1. should render the editexpensepage w/snapshot
// 2. should handle editExpense, using spies, trigger things, make sure spies got called
//3. Should handle removeExpense, similar to #2

let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage 
            startEditExpense={startEditExpense} 
            startRemoveExpense={startRemoveExpense} 
            expense={expenses[1]} 
            history={history} 
        />);
});

test('Should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle editExpense correctly', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('Should handle startRemoveExpense correctly', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
});