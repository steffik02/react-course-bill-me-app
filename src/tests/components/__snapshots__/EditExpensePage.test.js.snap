// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense page correctly 1`] = `
<div>
  <p>
    This is where you edit your expenses. I will be editing mine to ZERO!
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
